src/python/pants/backend/python/rules/pex.py

```diff
@@ -87,8 +162,6 @@ class Pex(HermeticPex):
     output_filename: str
 
 
-# TODO: This is non-hermetic because the requirements will be resolved on the fly by
```

Comment:
i'm not sure the TODO is done. We read constraints files now, but a constraint need not be a pin, it can be ranged.

Reply:
I'm unclear what there would be left to do, though? Is it possible to ever make requirements resolution be hermetic? Fundamentally, it's a very impure operation that depends on external state like what is available on PyPI. Constraint support seems like the closest thing we can get to hermicity.

Reply:
Hermeticity is achieved when we support generating lockfiles with hashes. This change is 1/2 the battle. It supports reading constraint files. Constraint files are looser than lock files; using them is really just a convenient hack. We could instead pass a lock file through as requirements.txt for example or have resolve just return a solution (urls + hashes) and have a separate download step. I'm not saying we should do any of these, just underscoring the convenient hack.

The remaining step to kill the TODO given the hack is writing constraint files with pinned versions and hashes.

Reply:
But I'm not sure that's something we can impose on repos. In other words, this lack of hermeticity in principle might be something we live with, as long as we can show users how to avoid it in practice by having a true lockfile.